src/server-glue/webresource-handler.ts

```diff
@@ -0,0 +1,344 @@
+			bb.removeAllListeners();
+		};
+
+		const clearFiles = () => {
+			isAborting = true;
+			const deletions = filesUploaded.map((file) => handler.removeFile(file));
+			// Best effort: We try to remove all uploaded files, but if this fails, there is not much to do
+			return Promise.all(deletions).catch((err) =>
+				console.error('Error deleting file', err),
+			);
+		};
+
+		bb.on('file', async (fieldname, filestream, info) => {
+			if (!isAborting && (await ifFileInValidPath(fieldname, req))) {
+				const file: IncomingFile = {
+					originalname: info.filename,
+					encoding: info.encoding,
+					mimetype: info.mimeType,
+					stream: filestream,
+					fieldname,
+				};
+				const promise = handler.handleFile(file).then((result) => {
+					req.body[fieldname] = {
+						filename: info.filename,
+						contentType: info.mimeType,
+						contentDisposition: undefined,
+						size: result.size,
+						href: result.filename,
+					};
+					filesUploaded.push(result.filename);
+				});
+				completeUploads.push(promise);
+			} else {
+				filestream.resume();
+			}
+		});
+
+		bb.on('field', (name, val, _info) => {
+			req.body[name] = val;
+		});
+
+		bb.on('finish', async () => {
+			try {
+				await Promise.all(completeUploads);
+				done();
+				next();
+			} catch (err: any) {
+				await clearFiles();
+
+				if (err instanceof FileSizeExceededError) {
+					return sbvrUtils.handleHttpErrors(
+						req,
+						res,
+						new errors.BadRequestError(err.message),
+					);
+				}
+
+				console.error('Error uploading file', err);
+				next(err);
+			}
+		});
+
+		bb.on('error', async (err) => {
+			await clearFiles();
+			done();
+			next(err);
+		});
+		req.pipe(bb);
+	};
+};
+
+// TODO: this can be cached
+const getWebResourceFields = (request: uriParser.ODataRequest): string[] => {
+	const fields =
+		request.abstractSqlModel?.tables[request.resourceName]?.modifyFields ??
+		request.abstractSqlModel?.tables[request.resourceName]?.fields;
+	if (fields == null) {
+		return [];
+	}
+
+	return fields
+		.filter((f) => f.dataType === 'WebResource')
+		.map((f) => sqlNameToODataName(f.fieldName));
+};
+
+const getModifiedFields = (request: uriParser.ODataRequest): string[] => {
+	return Object.entries(request.values)
+		.filter(([_key, value]) => value !== undefined)
+		.map(([key, _value]) => key);
+};
+
+const deleteFiles = async (
+	keysToDelete: string[],
+	webResourceHandler: WebResourceHandler,
+) => {
+	const promises = keysToDelete.map((r) => webResourceHandler.removeFile(r));
+	await Promise.all(promises);
+};
+
+const getCreateWebResourceHook = (webResourceHandler: WebResourceHandler) => {
+	return {
+		'POSTRUN-ERROR': async ({ tx, request }) => {
+			tx?.on('rollback', async () => {
+				const fields = getWebResourceFields(request);
+
+				if (fields.length === 0) {
+					return;
+				}
+
+				const keysToDelete: string[] = fields
+					.filter((f) => isDefined(request.values[f]))
+					.map((f) => request.values[f].href);
```

Comment:
I think this might flow better and have less lookups as:
```suggestion
					.map((f) => request.values[f]?.href);
					.filter(isDefined)
```

Reply:
@Page- That is a very nice catch, I think it was automatic to first filter and then map, map in this very specific case we can have less lookups as previously both operations have to do the request.values[f] access. Thanks!